Principles/6 - Declarative over imperative.md

```diff
@@ -0,0 +1,31 @@
+# Declarative over imperative
+
+Imperative code is often a natural choice for developers who have spent a lot of time in C-style languages. The primitives of most programming languages make it easy to write imperative code. However, building imperative APIs can lead to more error prone systems, more difficult to read code, and it being harder to discern the intention from the implementation for a given piece of logic.
+
+We prefer to write declarative code. This style allows us to define the relationships between data and UI and what results we would like to see, rather than the way we want them to happen. Where imperative code must exist a developer should attempt to add a more declarative wrapper encapsulating it as an implementation detail.
+
+Below are a few examples of decisions that demonstrate this principle:
+
+* We chose React over more traditional imperative DOM manipulation, or technologies like web components, for it's declarative API.
+
+* We chose GraphQL over REST for it's declarative query language and the declarative components provided by it's client libraries.
+
+* We prefer to render a component that encapsulates a series of imperative DOM actions, such as `<EventListener />` or `<ScrollLock />`, than to expose classes or functions that one would call directly.
+
+* We prefer to use or build declarative APIs on top of complex state heavy operations. Examples include managing form state with `<FormState />`, handling graphql requests and caching with `apollo-client`, and tracking long running jobs with `<JobPoller />`.
+
+* We prefer to build declarative components to express HTTP response behavior than to write imperative server code. For example; we built `<Redirect />` to replace imperative calls to `location.assign`.
+
+* We prefer to build components as pure functions when we can avoid writing a class. Classes can often result in more imperative code since they encourage manipulation of instance state.
+
+* We prefer declarative decorators for enhancing the behavior of class methods, such as `@autobind` or `@memoize`, over imperatively overwriting fields with wrapped versions or writing memoization logic by hand.
+
+* We prefer to express program logic through a series of `const` assignments rather than mutating a single variable.
+
+* We prefer to express list and object manipulation, through functional constructs (map/filter/reduce) than to build up a new list through a for loop.
+
+## Conflicts
+
```

Comment:
I’m not sure this section adds that much value

Reply:
I think it's important to highlight tensions between principles and times when we might want to make an exception